docs/architecture/adr-009-modular-antehandler.md

```diff
@@ -0,0 +1,173 @@
+
+A key benefit of this approach is that one Decorator can wrap its internal logic around the next Checker/Deliverer. A weave Decorator may do the following:
+
+```golang
+// Example Decorator's Deliver function
+func (example Decorator) Deliver(ctx Context, store KVStore, tx Tx, next Deliverer) {
+    // Do some pre-processing logic
+
+    res, err := next.Deliver(ctx, store, tx)
+
+    // Do some post-processing logic given the result and error
+}
+```
+
+Pros:
+1. Weave Decorators can wrap over the next decorator/handler in the chain. The ability to both pre-process and post-process may be useful in certain settings.
+2. Provides a nested modular structure that isn't possible in the solution above, while also allowing for a linear one-after-the-other structure like the solution above.
+
+Cons:
+1. It is hard to understand at first glance the state updates that would occur after a Decorator runs given the `ctx`, `store`, and `tx`. A Decorator can have an arbitrary number of nested Decorators being called within its function body, each possibly doing some pre- and post-processing before calling the next decorator on the chain. Thus to understand what a Decorator is doing, one must also understand what every other decorator further along the chain is also doing. This can get quite complicated to understand. A linear, one-after-the-other approach while less powerful, may be much easier to reason about.
+
+### Chained Micro-Functions
+
+The benefit of Weave's approach is that the Decorators can have very narrowly defined functionality, which when chained together allows for maximum customizability. However, the nested structure can get quite complex and thus hard to reason about.
+
+My preferred approach is to split the AnteHandler functionality into tightly scoped "micro-functions", while preserving the one-after-the-other ordering that would come from the per-module approach.
+
+We can then have a way to chain these micro-functions so that they run one after the other. Modules may define multiple ante micro-functions and then also provide a default per-module AnteHandler that implements a default, suggested order for these micro-functions. 
+
+Users can order the AnteHandlers easily by simply using the ModuleManager. The ModuleManager will take in a list of AnteHandlers and return a single AnteHandler that runs each AnteHandler in the order of the list provided. If the user is comfortable with the default ordering of each module, this is as simple as providing a list with each module's antehandler (exactly the same as BeginBlocker and EndBlocker).
+
+If however, users wish to change the order or add, modify, or delete ante micro-functions in anyway; they can always define their own ante micro-functions and add them explicitly to the list that gets passed into module manager.
```

Comment:
This is what I find to be better over decorators, especially if the chain of decorators in long @ethanfrey.

Reply:
>  If however, users wish to change the order or add, modify, or delete ante micro-functions in anyway; they can always define their own ante micro-functions and add them explicitly to the list that gets passed into module manager. 

How is this different than decorators?
They are identical in taking a list of functions. The same list.

The difference is that the decorator can run cleanup *after* the call, or modify the result.
It is the same chain of functions in both cases, just the power you give them. 

Thinks like making a sub-transaction after collecting fees (and only committing msg updates on success), or catching panics, are clear places where the chain design doesn't work. How do you plan to handle those?

Reply:
> If the decorator construction is done in a single place and not convoluted by abstraction, then the readability is the same in both approaches, I agree.

:+1:

I feel the `Module interface` is a very nice standard. But no need to force the `ModuleManager` on everything.